docs/how-to/push-update-on-mainnet.md

```diff
@@ -0,0 +1,118 @@
+Overview
+========
+
+Software update is a protocol mechanism that allows nodes to
+automatically update their executables without changing protocol
+itself. Here's the process overview:
+1. Software update is proposed. Proposal is a datatype that gets into
+   the blockchain. It contains information about version changes and
+   hashes of update files.
+2. Update files are uploaded to the S3 bucket.
+3. Software update is confirmed by voting from majority of nodes.
+4. Nodes that see an update try to download and apply it
+   automatically.
+
+Now, again, step-by-step.
+
+Prerequisites
+=============
+1. Cluster nodes' keys. They are crucial to have since cluster nodes
+   are the only nodes that have stake in the system so they are the
+   only nodes that can propose an update and vote for it.
+2. Access to the S3 bucket
+3. Installers corresponding to software version we're pushing to cluster
+   - Those are two executables (exe for win, pkg for mac) which are to be provided by QA (after QA procedures passed)
+   - Advice: after providing installers, ask QA to confirm hashes are same, we don't want to ship users incorrect installers to users
+4. Software version
+   - Single integer, denote application version as for installers provided
+
+Proposing the update
+====================
+
+Ensure your keys are somewhere around and named, say, `nodeX.key`,
+where `X ∈ [1..7]`. Procedure is done using `cardano-auxx`
+executable.
+
+You should change param `--peer` for all commands listed below. Also consider changing `--log-config`, `--logs-prefix`.
+Note `--system-start 0` is perfectly valid because we don't need actual value for `cardano-auxx`.
+
+Import secret keys (only 4 of 7 needed, because we need majority):
+
+```
+stack exec cardano-auxx -- --system-start 0 --log-config
+  scripts/log-templates/log-config-qa.yaml
+  --logs-prefix "logs/auxx-update-1.0.1" --db-path auxx-update-1.0.1
+  --peer <relay dns name>:3000 cmd --commands "add-key node1.key, add-key node2.key, add-key node3.key, add-key node4.key, listaddr"
+```
```

Comment:
This combination of  "import" and "propose-update" below suggests that keys are actually copied somewhere by this procedure.

This sounds troublesome, and raises a question -- has this procedure undergone security audit?

Reply:
In case the keys are indeed copied to another location -- perhaps the update mechanism could be reimplemented, so that there is no need for that?

Reply:
Keys are imported into the local wallet secret key file `secret.key` right in the folder you execute `auxx` in.

Reply:
@volhovM, how does the cluster authenticate the keys without their contents being copied? Is the authentication done over a secure channel?

Or maybe I have this all wrong, and should think of the node keys signing some asset to indicate their intent to propose an update.

Reply:
> Or maybe I have this all wrong, and should think of the node keys signing some asset to indicate their intent to propose an update.

That's exactly right. The node keys sign an update proposal, and then it's broadcast to other nodes as a special message (similarly to how transactions are sent) and ultimately included into the next block.

Reply:
@volhovM so I think we should document here the fact that node keys are imported into the local wallet db.

Reply:
Done.